// @trimble-oss/modus-web-components/dist/collection/components/modus-table/parts/row/actions/modus-table-row-actions-cell.js
import { Host, h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from "@stencil/core";
import NavigateTableCells from "../../../utilities/table-cell-navigation.utility";
import { KEYBOARD_ENTER } from "../../../modus-table.constants";
export class ModusTableRowActionsCell {
    constructor() {
        this.onCellKeyDown = (e) => this.handleCellKeydown(e);
        this.handleCellKeydown = (event) => {
            var _a, _b, _c;
            if (event.defaultPrevented)
                return;
            const key = (_a = event.key) === null || _a === void 0 ? void 0 : _a.toLowerCase();
            if (key === KEYBOARD_ENTER) {
                (_c = (_b = this.el.firstChild) === null || _b === void 0 ? void 0 : _b.firstChild) === null || _c === void 0 ? void 0 : _c.focusButton();
                event.stopPropagation();
            }
            else {
                NavigateTableCells({
                    eventKey: event.key,
                    cellElement: this.cellEl,
                });
            }
        };
        this.row = undefined;
        this.context = undefined;
    }
    connectedCallback() {
        this.cellEl = this.el.parentElement;
        this.cellEl.addEventListener('keydown', this.onCellKeyDown);
    }
    disconnectedCallback() {
        if (this.cellEl) {
            this.cellEl.removeEventListener('keydown', this.onCellKeyDown);
        }
    }
    render() {
        return (h(Host, { key: 'a7c39d0729d1b511ce0b50e1a2a0c4530bb2be8a' }, h("modus-table-row-actions", { key: '18eb35d80a7917256ea02074fe46dac4a4383f3a', row: this.row, context: this.context })));
    }
    static get is() { return "modus-table-row-actions-cell"; }
    static get properties() {
        return {
            "row": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Row<unknown>",
                    "resolved": "Row<unknown>",
                    "references": {
                        "Row": {
                            "location": "import",
                            "path": "@tanstack/table-core",
                            "id": ""
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            },
            "context": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TableContext",
                    "resolved": "TableContext",
                    "references": {
                        "TableContext": {
                            "location": "import",
                            "path": "../../../models/table-context.models",
                            "id": "src/components/modus-table/models/table-context.models.ts::TableContext"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            }
        };
    }
    static get elementRef() { return "el"; }
}
