// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/profile-menu/modus-navbar-profile-menu.js
// eslint-disable-next-line
import { h } from "@stencil/core";
import { ModusIconMap } from "../../../icons/ModusIconMap";
export class ModusNavbarProfileMenu {
    constructor() {
        this.avatarUrl = undefined;
        this.email = undefined;
        this.initials = undefined;
        this.links = undefined;
        this.signOutText = 'Sign out';
        this.reverse = undefined;
        this.username = undefined;
        this.variant = 'default';
    }
    signOutKeydownHandler(event) {
        if (event.code !== 'Enter') {
            return;
        }
        this.signOutClick.emit();
    }
    render() {
        var _a;
        const direction = this.reverse ? 'reverse' : '';
        const variant = this.variant === 'default' ? '' : `profile-menu-${this.variant}`;
        return (h("div", { key: '950336bb8022523f5d9d82b680b1f93627394c66', class: `profile-menu ${direction} ${variant}`, onClick: (event) => event.preventDefault() }, h("div", { key: '3ade3406c483372cdd6ff51e8ec076c3de2a0abe', class: "user" }, this.avatarUrl ? (h("img", { class: "avatar", src: this.avatarUrl, alt: "Avatar" })) : (h("span", { class: "initials" }, this.initials)), h("div", { key: '3fab86cf79860d4d8a450d67657747a2ff980469', class: "user-details" }, h("div", { key: 'c6bbb2a6a57ec589696042009fc681907852e693', class: "username" }, this.username), h("div", { key: 'e7badf44abd0124cd3e3ff7632e878212a9bdb97', class: "email" }, this.email))), ((_a = this.links) === null || _a === void 0 ? void 0 : _a.length) ? (h("modus-list", { class: "links" }, this.links.map((link) => {
            return (h("modus-list-item", { borderless: true, onItemClick: () => this.linkClick.emit(link.id) }, h("div", { class: "link-item" }, link.icon && (h("div", { class: "icon" }, h(ModusIconMap, { icon: link.icon, size: "24" }))), link.display)));
        }))) : null, h("slot", { key: '4056c5fcd4b8180ce1db6424b4b6a28657ba3497' }), h("div", { key: '85a779b160a7cba9e63053d122a363e70e19f721', class: "sign-out", onClick: () => this.signOutClick.emit(), onKeyDown: (event) => this.signOutKeydownHandler(event), tabIndex: 0 }, h("div", { key: '1d7400ca1ae72f7b056f93ac95daf283461badd5' }, this.signOutText))));
    }
    static get is() { return "modus-navbar-profile-menu"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-navbar-profile-menu.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-navbar-profile-menu.css"]
        };
    }
    static get properties() {
        return {
            "avatarUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "avatar-url",
                "reflect": false
            },
            "email": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "email",
                "reflect": false
            },
            "initials": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "initials",
                "reflect": false
            },
            "links": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ModusNavbarProfileMenuLink[]",
                    "resolved": "ModusNavbarProfileMenuLink[]",
                    "references": {
                        "ModusNavbarProfileMenuLink": {
                            "location": "import",
                            "path": "../modus-navbar.models",
                            "id": "src/components/modus-navbar/modus-navbar.models.ts::ModusNavbarProfileMenuLink"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            },
            "signOutText": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "sign-out-text",
                "reflect": false,
                "defaultValue": "'Sign out'"
            },
            "reverse": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "reverse",
                "reflect": false
            },
            "username": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "username",
                "reflect": false
            },
            "variant": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'default' | 'blue'",
                    "resolved": "\"blue\" | \"default\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "variant",
                "reflect": false,
                "defaultValue": "'default'"
            }
        };
    }
    static get events() {
        return [{
                "method": "linkClick",
                "name": "linkClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                }
            }, {
                "method": "signOutClick",
                "name": "signOutClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "MouseEvent",
                    "resolved": "MouseEvent",
                    "references": {
                        "MouseEvent": {
                            "location": "global",
                            "id": "global::MouseEvent"
                        }
                    }
                }
            }];
    }
}
