// @trimble-oss/modus-web-components/dist/collection/components/modus-navbar/apps-menu/modus-navbar-apps-menu.js
// eslint-disable-next-line
import { h } from "@stencil/core";
export class ModusNavbarAppsMenu {
    constructor() {
        this.apps = undefined;
        this.reverse = undefined;
    }
    clickAppHandler(event, app) {
        event.preventDefault();
        window.open(app.url, '_blank');
        this.appOpen.emit(app);
    }
    render() {
        var _a;
        const direction = this.reverse ? 'reverse' : '';
        return (h("div", { key: '43839464e5f074c69989af7647bc6a228cd8d150', class: `apps-menu ${direction}`, onClick: (event) => event.preventDefault() }, (_a = this.apps) === null || _a === void 0 ? void 0 : _a.map((app) => (h("div", { class: "app-div" }, app.showCategory ? h("div", { class: "category" }, app.category) : null, h("a", { class: "app", href: app.url, onClick: (event) => this.clickAppHandler(event, app) }, h("img", { src: app.logoUrl, alt: "", height: "32", width: "32" }), h("div", { class: "right" }, h("div", { class: "name" }, app.name), app.description ? h("div", { class: "description" }, app.description) : null)))))));
    }
    static get is() { return "modus-navbar-apps-menu"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-navbar-apps-menu.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-navbar-apps-menu.css"]
        };
    }
    static get properties() {
        return {
            "apps": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "ModusNavbarApp[]",
                    "resolved": "ModusNavbarApp[]",
                    "references": {
                        "ModusNavbarApp": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-navbar/apps-menu/modus-navbar-apps-menu.tsx",
                            "id": "src/components/modus-navbar/apps-menu/modus-navbar-apps-menu.tsx::ModusNavbarApp"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                }
            },
            "reverse": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "attribute": "reverse",
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "appOpen",
                "name": "appOpen",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": ""
                },
                "complexType": {
                    "original": "ModusNavbarApp",
                    "resolved": "ModusNavbarApp",
                    "references": {
                        "ModusNavbarApp": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-navbar/apps-menu/modus-navbar-apps-menu.tsx",
                            "id": "src/components/modus-navbar/apps-menu/modus-navbar-apps-menu.tsx::ModusNavbarApp"
                        }
                    }
                }
            }];
    }
}
