// @trimble-oss/modus-web-components/dist/collection/components/modus-message/modus-message.js
// eslint-disable-next-line
import { h } from "@stencil/core";
import { IconInfo } from "../../icons/generated-icons/IconInfo";
import { IconHelp } from "../../icons/generated-icons/IconHelp";
import { ModusIconMap } from "../../icons/ModusIconMap";
export class ModusMessage {
    constructor() {
        this.classByType = new Map([
            ['info', 'info'],
            ['question', 'question'],
        ]);
        this.ariaLabel = undefined;
        this.icon = undefined;
        this.type = 'info';
    }
    render() {
        const className = `modus-message ${this.classByType.get(this.type)}`;
        return (h("div", { key: 'dd6016df199cc6bdec85e3ed234ecf6caac71270', "aria-label": this.ariaLabel || undefined, class: className, role: "note" }, h("span", { key: 'd5fdb22343015cde0bbee4310a6ea79ef4962c43', class: "icon" }, this.icon ? (h(ModusIconMap, { icon: this.icon, size: "18" })) : this.type === 'info' ? (h(IconInfo, { size: "18" })) : this.type === 'question' ? (h(IconHelp, { size: "18" })) : null), h("span", { key: 'c2295954d69d9660192c6094d6751321845bc0ab', class: "message" }, h("slot", { key: 'c615fc24a81859e9ac8fe450468f03d4cb5fd608' }))));
    }
    static get is() { return "modus-message"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-message.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-message.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The message's aria-label."
                },
                "attribute": "aria-label",
                "reflect": false
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "(optional) The message's icon."
                },
                "attribute": "icon",
                "reflect": false
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'info' | 'question'",
                    "resolved": "\"info\" | \"question\"",
                    "references": {}
                },
                "required": false,
                "optional": true,
                "docs": {
                    "tags": [],
                    "text": "(optional) The message's type."
                },
                "attribute": "type",
                "reflect": false,
                "defaultValue": "'info'"
            }
        };
    }
}
