// @trimble-oss/modus-web-components/dist/collection/components/modus-chip/modus-chip.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { h } from "@stencil/core";
import { IconRemove } from "../../icons/svgs/icon-remove";
import { IconCheck } from "../../icons/svgs/icon-check";
export class ModusChip {
    constructor() {
        this.classByChipStyle = new Map([
            ['outline', 'style-outline'],
            ['solid', 'style-solid'],
        ]);
        this.classBySize = new Map([
            ['medium', 'medium'],
            ['small', 'small'],
        ]);
        this.ariaLabel = undefined;
        this.chipStyle = 'solid';
        this.disabled = false;
        this.hasError = false;
        this.imageUrl = undefined;
        this.showCheckmark = false;
        this.showClose = false;
        this.size = 'medium';
        this.value = undefined;
        this.chipId = undefined;
        this.maxWidth = undefined;
        this.active = false;
    }
    elementKeyupHandler(event) {
        switch (event.code) {
            case 'Escape':
                if (!this.showClose) {
                    return;
                }
                this.closeClick.emit(event);
                break;
        }
    }
    elementKeydownHandler(event) {
        switch (event.code) {
            case 'Enter':
                this.chipClick.emit(event);
                break;
        }
    }
    onChipClick(event) {
        if (event.defaultPrevented) {
            return;
        } // Don't emit chipClick if closeClick has emitted.
        this.chipClick.emit(event);
    }
    onCloseClick(event) {
        this.closeClick.emit(event);
        event.preventDefault();
    }
    render() {
        var _a;
        const chipClass = `
      modus-chip ${this.disabled ? 'disabled' : ''}
      ${this.hasError ? 'has-error' : ''}
      ${this.classByChipStyle.get(this.chipStyle)}
      ${this.classBySize.get(this.size)}
      ${!this.showCheckmark && !this.imageUrl ? 'no-left-icon' : null}
      ${!this.showClose ? 'no-right-icon' : null}
      ${this.active ? 'active' : ''}
    `;
        const style = {
            style: {
                'max-width': (_a = this.maxWidth) !== null && _a !== void 0 ? _a : undefined,
            },
        };
        return (h("button", { key: 'a2b1f4dd79429b8819c3d86cb15ec83e88798134', "aria-disabled": this.disabled ? 'true' : undefined, "aria-label": this.ariaLabel || undefined, id: this.chipId || undefined, class: chipClass, onClick: this.disabled ? null : (event) => this.onChipClick(event), tabIndex: 0, type: "button" }, this.imageUrl ? (h("img", { src: this.imageUrl, alt: "" })) : this.showCheckmark ? (h(IconCheck, { size: '16' })) : null, h("span", Object.assign({ key: '5f096da0d11883a39dca200ec7b97d08762fc230' }, style), this.value), this.showClose ? (h(IconRemove, { onClick: this.disabled ? null : (event) => this.onCloseClick(event), size: '16' })) : null));
    }
    static get is() { return "modus-chip"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-chip.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-chip.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The chip's aria-label."
                },
                "attribute": "aria-label",
                "reflect": false
            },
            "chipStyle": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'outline' | 'solid'",
                    "resolved": "\"outline\" | \"solid\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The chip's style."
                },
                "attribute": "chip-style",
                "reflect": false,
                "defaultValue": "'solid'"
            },
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether the chip is disabled."
                },
                "attribute": "disabled",
                "reflect": false,
                "defaultValue": "false"
            },
            "hasError": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether the chip has an error."
                },
                "attribute": "has-error",
                "reflect": false,
                "defaultValue": "false"
            },
            "imageUrl": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The image's url."
                },
                "attribute": "image-url",
                "reflect": false
            },
            "showCheckmark": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether to show the checkmark."
                },
                "attribute": "show-checkmark",
                "reflect": false,
                "defaultValue": "false"
            },
            "showClose": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether to show the close icon."
                },
                "attribute": "show-close",
                "reflect": false,
                "defaultValue": "false"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'medium' | 'small'",
                    "resolved": "\"medium\" | \"small\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The chip's size."
                },
                "attribute": "size",
                "reflect": false,
                "defaultValue": "'medium'"
            },
            "value": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The chip's value."
                },
                "attribute": "value",
                "reflect": false
            },
            "chipId": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) the chip's id"
                },
                "attribute": "chip-id",
                "reflect": false
            },
            "maxWidth": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Maximum width for the Chip's text and shows ellipsis when truncated"
                },
                "attribute": "max-width",
                "reflect": false
            },
            "active": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether the chip is active."
                },
                "attribute": "active",
                "reflect": false,
                "defaultValue": "false"
            }
        };
    }
    static get events() {
        return [{
                "method": "chipClick",
                "name": "chipClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires on chip click."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "closeClick",
                "name": "closeClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires on close icon click."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
    static get listeners() {
        return [{
                "name": "keyup",
                "method": "elementKeyupHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }, {
                "name": "keydown",
                "method": "elementKeydownHandler",
                "target": undefined,
                "capture": false,
                "passive": false
            }];
    }
}
