// @trimble-oss/modus-web-components/dist/collection/components/modus-table/parts/panel/modus-table-toolbar/modus-table-toolbar.js
import { Host, h, // eslint-disable-line @typescript-eslint/no-unused-vars
 } from "@stencil/core";
export class ModusTablePanel {
    constructor() {
        this.context = undefined;
    }
    render() {
        // const { tableInstance: table, toolbarOptions: options } = this.context;
        return (h(Host, { key: 'fcbd8a1bec08ca5f4a11de68eac6b2a708127ee1' }, h("div", { key: 'd218fafadd7beb53e8c43ab0e45bf566a5996a0b', class: "table-toolbar" }, h("div", { key: 'e1d07736405cbb3828070c2f9b93a87fb9848937', class: "section" }, h("slot", { key: '89737c466d0a8e0baea9a4e9f89bf64893bc110c', name: "group-left" })), h("div", { key: '86342197962d9c921c8b8421539cee22a36226f9', class: "section" }, h("slot", { key: 'eb7745a45dc1c629d174414004efb47803971b49', name: "group-right" }), h("modus-table-dropdown-menu", { context: this.context })))));
    }
    static get is() { return "modus-table-toolbar"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["./modus-table-toolbar.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-table-toolbar.css"]
        };
    }
    static get properties() {
        return {
            "context": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "TableContext",
                    "resolved": "TableContext",
                    "references": {
                        "TableContext": {
                            "location": "import",
                            "path": "../../../models/table-context.models",
                            "id": "src/components/modus-table/models/table-context.models.ts::TableContext"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "Table data."
                }
            }
        };
    }
}
