// @trimble-oss/modus-web-components/dist/collection/components/modus-accordion/modus-accordion.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { h } from "@stencil/core";
export class ModusAccordion {
    constructor() {
        this.ariaLabel = undefined;
    }
    render() {
        return (h("div", { key: 'c3e49155ddc4b69b7ac12d3c679c814605e86858', "aria-label": this.ariaLabel || undefined, class: "accordion" }, h("slot", { key: 'd22c2cd38bae0126dbbedd6024c127d951cf4727' })));
    }
    static get is() { return "modus-accordion"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-accordion.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-accordion.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The accordion's aria-label."
                },
                "attribute": "aria-label",
                "reflect": false
            }
        };
    }
}
