// @trimble-oss/modus-web-components/dist/collection/components/modus-accordion-item/modus-accordion-item.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { h } from "@stencil/core";
import { generateElementId } from "../../utils/utils";
import { IconExpandMoreCircle } from "../../icons/generated-icons/IconExpandMoreCircle";
import { IconExpandMore } from "../../icons/generated-icons/IconExpandMore";
import { ModusIconMap } from "../../icons/ModusIconMap";
export class ModusAccordionItem {
    constructor() {
        this.expandedContentId = generateElementId() + '_accordion-item';
        this.classBySize = new Map([
            ['condensed', 'small'],
            ['standard', 'standard'],
        ]);
        // Trick to restart an element's animation
        // see https://www.charistheo.io/blog/2021/02/restart-a-css-animation-with-javascript/#restarting-a-css-animation
        // taken from: https://getbootstrap.com/docs/5.3/dist/js/bootstrap.js
        this.reflow = (element) => {
            element.offsetHeight; // eslint-disable-line no-unused-expressions
        };
        this.disabled = undefined;
        this.expandButtonType = 'standardArrow';
        this.expanded = undefined;
        this.headerText = undefined;
        this.icon = undefined;
        this.size = 'standard';
    }
    handleHeaderClick() {
        if (this.disabled) {
            return;
        }
        this.chevronContainerRef.classList.toggle('reverse');
        // Logic to trigger a transition animation and handle it
        // Because transition won't have any effect when using 'display:' on an element
        if (!this.expanded) {
            this.accordionBodyRef.classList.remove('collapse');
            this.accordionBodyRef.classList.add('collapsing');
            // Required to calculate scrollHeight and set the value on 'height' for transition to start
            this.accordionBodyRef.style.height = '0';
            // Timeout to reset collapsing class
            this.accordionOpenTimeout = setTimeout(() => {
                this.accordionBodyRef.classList.remove('collapsing');
                this.accordionBodyRef.classList.add('show');
                this.accordionBodyRef.classList.add('collapse');
                // reset height to original state
                // this.accordionBodyRef.style.height = '';
                clearTimeout(this.accordionOpenTimeout);
                this.expanded = true;
                this.opened.emit();
            }, 350);
            // Triggers transition
            this.accordionBodyRef.style.height = `${this.accordionBodyRef.scrollHeight}px`;
        }
        else {
            this.accordionBodyRef.style.height = `${this.accordionBodyRef.getBoundingClientRect().height}px`;
            this.reflow(this.accordionBodyRef);
            this.accordionBodyRef.classList.add('collapsing');
            this.accordionBodyRef.classList.remove('collapse');
            this.accordionBodyRef.classList.remove('show');
            // Timeout to reset collapsing class
            this.accordionCloseTimeout = setTimeout(() => {
                this.accordionBodyRef.classList.remove('collapsing');
                this.accordionBodyRef.classList.add('collapse');
                clearTimeout(this.accordionCloseTimeout);
                this.expanded = false;
                this.closed.emit();
            }, 350);
            this.accordionBodyRef.style.height = '';
        }
    }
    handleKeydown(event) {
        if (event.code !== 'Enter') {
            return;
        }
        this.handleHeaderClick();
    }
    renderIcon() {
        return (h("span", { class: "icon" }, h(ModusIconMap, { icon: this.icon })));
    }
    render() {
        const sizeClass = `${this.classBySize.get(this.size)}`;
        const disabledClass = `${this.disabled ? 'disabled' : ''}`;
        const expandedClass = `${this.expanded ? 'expanded' : ''}`;
        const bodyClass = `body ${sizeClass} collapse${this.expanded ? ' show' : ''}`;
        const headerClass = `header ${sizeClass} ${disabledClass} ${expandedClass}`;
        return (h("div", { key: 'e9f9be8a4b01ce99791b4c0dee521d09a1566fa1', "aria-disabled": this.disabled ? 'true' : undefined, "aria-expanded": this.expanded ? 'true' : undefined, class: "accordion-item" }, h("div", { key: '38a8ede0acd1b7fbeba51e978043546e759ef5f3', class: headerClass, role: "button", "aria-expanded": this.expanded ? 'true' : 'false', "aria-controls": this.expandedContentId, onClick: () => this.handleHeaderClick(), onKeyDown: (event) => this.handleKeydown(event), tabIndex: this.disabled ? -1 : 0 }, this.icon ? this.renderIcon() : null, h("span", { key: 'e0d8001a9b079cfa636399c8551bcfb3de0a390c', class: "title" }, this.headerText), h("div", { class: `chevron-container ${this.expanded ? 'reverse' : ''} `, ref: (el) => (this.chevronContainerRef = el) }, this.expandButtonType == 'circleArrow' ? (h(IconExpandMoreCircle, { size: "24" })) : (h(IconExpandMore, { size: "24" })))), h("div", { key: 'de48ff60b9911986e912528d85ae99511a14c0cf', id: this.expandedContentId, class: bodyClass, ref: (el) => (this.accordionBodyRef = el) }, h("div", { key: '825eca44853644c0070f10bf4c3252a25222e32f', class: "body-content" }, h("slot", { key: 'a8b5ca36212184a2f85e650cd40ad5baafa47358' })))));
    }
    static get is() { return "modus-accordion-item"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-accordion-item.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-accordion-item.css"]
        };
    }
    static get properties() {
        return {
            "disabled": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Disables the accordion item, locks expand/collapse."
                },
                "attribute": "disabled",
                "reflect": false
            },
            "expandButtonType": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'standardArrow' | 'circleArrow'",
                    "resolved": "\"circleArrow\" | \"standardArrow\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The type of expand button"
                },
                "attribute": "expand-button-type",
                "reflect": false,
                "defaultValue": "'standardArrow'"
            },
            "expanded": {
                "type": "boolean",
                "mutable": true,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) Whether the accordion item is expanded."
                },
                "attribute": "expanded",
                "reflect": false
            },
            "headerText": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(required) The text to render in the header."
                },
                "attribute": "header-text",
                "reflect": false
            },
            "icon": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The icon to display before the header text."
                },
                "attribute": "icon",
                "reflect": false
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "'condensed' | 'standard'",
                    "resolved": "\"condensed\" | \"standard\"",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The size of accordion item."
                },
                "attribute": "size",
                "reflect": false,
                "defaultValue": "'standard'"
            }
        };
    }
    static get events() {
        return [{
                "method": "closed",
                "name": "closed",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires on every accordion close."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }, {
                "method": "opened",
                "name": "opened",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "An event that fires on every accordion open."
                },
                "complexType": {
                    "original": "any",
                    "resolved": "any",
                    "references": {}
                }
            }];
    }
}
