// @trimble-oss/modus-web-components/dist/cjs/modus-breadcrumb.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-4b22c494.js');

const modusBreadcrumbCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");nav{font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";font-size:0.875rem;position:relative}nav ol{align-items:center;display:flex;flex-wrap:wrap;list-style:none;margin:0;padding:0.625rem 0}nav ol li .crumb a{color:var(--modus-link-color, #217cbb);cursor:pointer;text-decoration:none}nav ol li .crumb a:hover{text-decoration:underline}nav ol li .crumb a:active{color:var(--modus-breadcrumb-item-active-color, #6c757d);text-decoration:none}nav ol li .crumb.underline a{text-decoration:underline}nav ol li .crumb .divider{color:var(--modus-breadcrumb-divider-color, #6c757d);cursor:default;padding:0 0.5rem}nav ol li .last-crumb{color:var(--modus-breadcrumb-item-active-color, #6c757d)}";
const ModusBreadcrumbStyle0 = modusBreadcrumbCss;

const ModusBreadcrumb = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.crumbClick = index.createEvent(this, "crumbClick", 7);
        this.ariaLabel = undefined;
        this.crumbs = [];
        this.underlineLinks = undefined;
    }
    render() {
        return (index.h("nav", { key: 'dd4973ca7376fe9f3a407c1b2b2a7df5aa4dee03', "aria-label": this.ariaLabel || undefined }, index.h("ol", { key: 'c4f4f45b97aaccef1d87ce26d9fa55621416d386' }, this.crumbs.map((crumb, index$1) => (index.h("li", { key: crumb.id }, index$1 < this.crumbs.length - 1 ? (index.h("span", { class: `crumb ${this.underlineLinks ? 'underline' : ''}` }, index.h("a", { onClick: () => this.crumbClick.emit(crumb) }, crumb.display), index.h("span", { class: "divider" }, '>'))) : (index.h("span", { class: "last-crumb", "aria-current": "page" }, crumb.display))))))));
    }
};
ModusBreadcrumb.style = ModusBreadcrumbStyle0;

exports.modus_breadcrumb = ModusBreadcrumb;
