// @trimble-oss/modus-web-components/dist/cjs/modus-table.constants-b3626e38.js
'use strict';

const SORTED_ASCENDING = 'Sorted Ascending';
const SORTED_DESCENDING = 'Sorted Descending';
const SORT_ASCENDING = 'Sort Ascending';
const SORT_DESCENDING = 'Sort Descending';
const HTML_ATTR_DATA_ACCESSOR_KEY = 'data-accessor-key';
const COLUMN_DEF_DATATYPE_KEY = 'dataType';
const COLUMN_DEF_SUB_ROWS_KEY = 'subRows';
const COLUMN_DEF_CELL_EDITOR_TYPE_KEY = 'cellEditorType';
const COLUMN_DEF_CELL_EDITOR_ARGS_KEY = 'cellEditorArgs';
const COLUMN_DEF_SHOWTOTAL = 'showTotal';
const COLUMN_DEF_ROW_SELECTION_ID = 'row-checkbox';
const COLUMN_DEF_ROW_SELECTION_CSS = 'row-checkbox';
const COLUMN_DEF_DATATYPE_TEXT = 'text';
const COLUMN_DEF_DATATYPE_INTEGER = 'integer';
const COLUMN_DEF_DATATYPE_LINK = 'link';
const COLUMN_DEF_DATATYPE_DATE = 'date';
const COLUMN_DEF_DATATYPE_BADGE = 'badge';
const CELL_EDIT_TYPE_TEXT = COLUMN_DEF_DATATYPE_TEXT;
const CELL_EDIT_TYPE_INT = COLUMN_DEF_DATATYPE_INTEGER;
const CELL_EDIT_TYPE_SELECT = 'select';
const CELL_EDIT_TYPE_AUTOCOMPLETE = 'autocomplete';
const CELL_EDIT_TYPE_DATE = COLUMN_DEF_DATATYPE_DATE;
const ALLOWED_CELL_EDIT_TYPES = [
    CELL_EDIT_TYPE_TEXT,
    CELL_EDIT_TYPE_INT,
    CELL_EDIT_TYPE_SELECT,
    CELL_EDIT_TYPE_AUTOCOMPLETE,
    CELL_EDIT_TYPE_DATE,
];
const PAGINATION_DEFAULT_SIZES = [10, 20, 30];
const PAGINATION_PAGEVIEW_TEXT = 'Page View';
const PAGINATION_SUMMARY_TEXT = 'Showing result';
// Keyboard key constants
const KEYBOARD_ENTER = 'enter';
const KEYBOARD_TAB = 'tab';
const KEYBOARD_LEFT = 'arrowleft';
const KEYBOARD_RIGHT = 'arrowright';
const KEYBOARD_UP = 'arrowup';
const KEYBOARD_DOWN = 'arrowdown';
const KEYBOARD_SPACE = ' ';
const KEYBOARD_ESCAPE = 'escape';
const EXPANDED_STATE_KEY = 'expanded';
const SORTING_STATE_KEY = 'sorting';
const ROW_SELECTION_STATE_KEY = 'rowSelection';
const COLUMN_SIZING_INFO_STATE_KEY = 'columnSizingInfo';
const PAGINATION_STATE_KEY = 'pagination';
const COLUMN_SIZING_STATE_KEY = 'columnSizing';
const COLUMN_VISIBILITY_STATE_KEY = 'columnVisibility';
const COLUMN_ORDER_STATE_KEY = 'columnOrder';

exports.ALLOWED_CELL_EDIT_TYPES = ALLOWED_CELL_EDIT_TYPES;
exports.CELL_EDIT_TYPE_AUTOCOMPLETE = CELL_EDIT_TYPE_AUTOCOMPLETE;
exports.CELL_EDIT_TYPE_DATE = CELL_EDIT_TYPE_DATE;
exports.CELL_EDIT_TYPE_INT = CELL_EDIT_TYPE_INT;
exports.CELL_EDIT_TYPE_SELECT = CELL_EDIT_TYPE_SELECT;
exports.CELL_EDIT_TYPE_TEXT = CELL_EDIT_TYPE_TEXT;
exports.COLUMN_DEF_CELL_EDITOR_ARGS_KEY = COLUMN_DEF_CELL_EDITOR_ARGS_KEY;
exports.COLUMN_DEF_CELL_EDITOR_TYPE_KEY = COLUMN_DEF_CELL_EDITOR_TYPE_KEY;
exports.COLUMN_DEF_DATATYPE_BADGE = COLUMN_DEF_DATATYPE_BADGE;
exports.COLUMN_DEF_DATATYPE_INTEGER = COLUMN_DEF_DATATYPE_INTEGER;
exports.COLUMN_DEF_DATATYPE_KEY = COLUMN_DEF_DATATYPE_KEY;
exports.COLUMN_DEF_DATATYPE_LINK = COLUMN_DEF_DATATYPE_LINK;
exports.COLUMN_DEF_ROW_SELECTION_CSS = COLUMN_DEF_ROW_SELECTION_CSS;
exports.COLUMN_DEF_ROW_SELECTION_ID = COLUMN_DEF_ROW_SELECTION_ID;
exports.COLUMN_DEF_SHOWTOTAL = COLUMN_DEF_SHOWTOTAL;
exports.COLUMN_DEF_SUB_ROWS_KEY = COLUMN_DEF_SUB_ROWS_KEY;
exports.COLUMN_ORDER_STATE_KEY = COLUMN_ORDER_STATE_KEY;
exports.COLUMN_SIZING_INFO_STATE_KEY = COLUMN_SIZING_INFO_STATE_KEY;
exports.COLUMN_SIZING_STATE_KEY = COLUMN_SIZING_STATE_KEY;
exports.COLUMN_VISIBILITY_STATE_KEY = COLUMN_VISIBILITY_STATE_KEY;
exports.EXPANDED_STATE_KEY = EXPANDED_STATE_KEY;
exports.HTML_ATTR_DATA_ACCESSOR_KEY = HTML_ATTR_DATA_ACCESSOR_KEY;
exports.KEYBOARD_DOWN = KEYBOARD_DOWN;
exports.KEYBOARD_ENTER = KEYBOARD_ENTER;
exports.KEYBOARD_ESCAPE = KEYBOARD_ESCAPE;
exports.KEYBOARD_LEFT = KEYBOARD_LEFT;
exports.KEYBOARD_RIGHT = KEYBOARD_RIGHT;
exports.KEYBOARD_SPACE = KEYBOARD_SPACE;
exports.KEYBOARD_TAB = KEYBOARD_TAB;
exports.KEYBOARD_UP = KEYBOARD_UP;
exports.PAGINATION_DEFAULT_SIZES = PAGINATION_DEFAULT_SIZES;
exports.PAGINATION_PAGEVIEW_TEXT = PAGINATION_PAGEVIEW_TEXT;
exports.PAGINATION_STATE_KEY = PAGINATION_STATE_KEY;
exports.PAGINATION_SUMMARY_TEXT = PAGINATION_SUMMARY_TEXT;
exports.ROW_SELECTION_STATE_KEY = ROW_SELECTION_STATE_KEY;
exports.SORTED_ASCENDING = SORTED_ASCENDING;
exports.SORTED_DESCENDING = SORTED_DESCENDING;
exports.SORTING_STATE_KEY = SORTING_STATE_KEY;
exports.SORT_ASCENDING = SORT_ASCENDING;
exports.SORT_DESCENDING = SORT_DESCENDING;
