// @trimble-oss/modus-web-components/dist/cjs/modus-spinner.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-4b22c494.js');

const modusSpinnerCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");@keyframes spin{0%{transform:rotate(0deg)}100%{transform:rotate(360deg)}}div.spinner{animation:spin 0.75s linear infinite;border:0.25em solid #0063a3;border-radius:50%;border-right-color:transparent;display:inline-flex;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";forced-color-adjust:none}";
const ModusSpinnerStyle0 = modusSpinnerCss;

const ModusSpinner = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.variants = new Map([
            ['primary', '#0063a3'],
            ['secondary', '#6a6e79'],
            ['tertiary', '#cbcdd6'],
            ['dark', '#252a2e'],
            ['success', '#006638'],
            ['danger', '#da212c'],
            ['warning', '#e49325'],
        ]);
        this.color = 'primary';
        this.size = '2rem';
    }
    render() {
        const borderColor = this.variants.has(this.color)
            ? `var(--modus-${this.color}, ${this.variants.get(this.color)})`
            : this.color;
        return (index.h("div", { key: '356a23fd14627f54653952f359ae0f659e0f51a0', "aria-busy": "true", class: "spinner", style: {
                'border-color': borderColor,
                'border-right-color': 'transparent',
                height: this.size,
                width: this.size,
            } }));
    }
};
ModusSpinner.style = ModusSpinnerStyle0;

exports.modus_spinner = ModusSpinner;
