// @trimble-oss/modus-web-components/dist/cjs/modus-card.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-4b22c494.js');

const modusCardCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");article{background-color:var(--modus-card-bg, #fff);border:0.0625rem solid;border-color:transparent;border-radius:0.25rem;color:var(--modus-card-color, #252a2e);font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";overflow:hidden;position:relative}article.shadow:hover{box-shadow:0 0 4px var(--modus-card-shadow-color, rgba(54, 53, 69, 0.3019607843))}article.card-border{border-color:var(--modus-card-border-color, #e0e1e9)}";
const ModusCardStyle0 = modusCardCss;

const ModusCard = class {
    constructor(hostRef) {
        index.registerInstance(this, hostRef);
        this.ariaLabel = undefined;
        this.height = '269px';
        this.width = '240px';
        this.borderRadius = undefined;
        this.showCardBorder = true;
        this.showShadowOnHover = true;
    }
    render() {
        return (index.h("article", { key: '0178e74db22d082126c6f75e116f1fe6809332de', class: `${this.showShadowOnHover ? 'shadow' : ''} ${this.showCardBorder ? 'card-border' : ''}  `, "aria-label": this.ariaLabel || undefined, style: {
                height: this.height,
                width: this.width,
                'border-radius': this.borderRadius,
            } }, index.h("slot", { key: 'aee2eb69c0c3a4193eb9e9f5ab675f7ebfc826cf' })));
    }
};
ModusCard.style = ModusCardStyle0;

exports.modus_card = ModusCard;
