// @trimble-oss/modus-react-components/dist/components/stencil-generated/index.js
/* eslint-disable */
/* tslint:disable */
/* auto-generated react proxies */
import { createReactComponent } from './react-component-lib';
import { defineCustomElements } from '@trimble-oss/modus-web-components/loader';
defineCustomElements();
export const ModusAccordion = /*@__PURE__*/ createReactComponent('modus-accordion');
export const ModusAccordionItem = /*@__PURE__*/ createReactComponent('modus-accordion-item');
export const ModusActionBar = /*@__PURE__*/ createReactComponent('modus-action-bar');
export const ModusAlert = /*@__PURE__*/ createReactComponent('modus-alert');
export const ModusAutocomplete = /*@__PURE__*/ createReactComponent('modus-autocomplete');
export const ModusBadge = /*@__PURE__*/ createReactComponent('modus-badge');
export const ModusBreadcrumb = /*@__PURE__*/ createReactComponent('modus-breadcrumb');
export const ModusButton = /*@__PURE__*/ createReactComponent('modus-button');
export const ModusButtonGroup = /*@__PURE__*/ createReactComponent('modus-button-group');
export const ModusCard = /*@__PURE__*/ createReactComponent('modus-card');
export const ModusCheckbox = /*@__PURE__*/ createReactComponent('modus-checkbox');
export const ModusChip = /*@__PURE__*/ createReactComponent('modus-chip');
export const ModusDataTable = /*@__PURE__*/ createReactComponent('modus-data-table');
export const ModusDateInput = /*@__PURE__*/ createReactComponent('modus-date-input');
export const ModusDatePicker = /*@__PURE__*/ createReactComponent('modus-date-picker');
export const ModusDivider = /*@__PURE__*/ createReactComponent('modus-divider');
export const ModusDropdown = /*@__PURE__*/ createReactComponent('modus-dropdown');
export const ModusFileDropzone = /*@__PURE__*/ createReactComponent('modus-file-dropzone');
export const ModusIcon = /*@__PURE__*/ createReactComponent('modus-icon');
export const ModusList = /*@__PURE__*/ createReactComponent('modus-list');
export const ModusListItem = /*@__PURE__*/ createReactComponent('modus-list-item');
export const ModusMessage = /*@__PURE__*/ createReactComponent('modus-message');
export const ModusModal = /*@__PURE__*/ createReactComponent('modus-modal');
export const ModusNavbar = /*@__PURE__*/ createReactComponent('modus-navbar');
export const ModusNavbarAppsMenu = /*@__PURE__*/ createReactComponent('modus-navbar-apps-menu');
export const ModusNavbarButtonMenu = /*@__PURE__*/ createReactComponent('modus-navbar-button-menu');
export const ModusNavbarMainMenu = /*@__PURE__*/ createReactComponent('modus-navbar-main-menu');
export const ModusNavbarNotificationsMenu = /*@__PURE__*/ createReactComponent('modus-navbar-notifications-menu');
export const ModusNavbarProfileMenu = /*@__PURE__*/ createReactComponent('modus-navbar-profile-menu');
export const ModusNavbarSearchOverlay = /*@__PURE__*/ createReactComponent('modus-navbar-search-overlay');
export const ModusNumberInput = /*@__PURE__*/ createReactComponent('modus-number-input');
export const ModusPagination = /*@__PURE__*/ createReactComponent('modus-pagination');
export const ModusProgressBar = /*@__PURE__*/ createReactComponent('modus-progress-bar');
export const ModusRadioGroup = /*@__PURE__*/ createReactComponent('modus-radio-group');
export const ModusSelect = /*@__PURE__*/ createReactComponent('modus-select');
export const ModusSentimentScale = /*@__PURE__*/ createReactComponent('modus-sentiment-scale');
export const ModusSideNavigation = /*@__PURE__*/ createReactComponent('modus-side-navigation');
export const ModusSideNavigationItem = /*@__PURE__*/ createReactComponent('modus-side-navigation-item');
export const ModusSlider = /*@__PURE__*/ createReactComponent('modus-slider');
export const ModusSpinner = /*@__PURE__*/ createReactComponent('modus-spinner');
export const ModusSwitch = /*@__PURE__*/ createReactComponent('modus-switch');
export const ModusTable = /*@__PURE__*/ createReactComponent('modus-table');
export const ModusTableCellEditor = /*@__PURE__*/ createReactComponent('modus-table-cell-editor');
export const ModusTableCellMain = /*@__PURE__*/ createReactComponent('modus-table-cell-main');
export const ModusTableColumnsVisibility = /*@__PURE__*/ createReactComponent('modus-table-columns-visibility');
export const ModusTableDropdownMenu = /*@__PURE__*/ createReactComponent('modus-table-dropdown-menu');
export const ModusTableFillerColumn = /*@__PURE__*/ createReactComponent('modus-table-filler-column');
export const ModusTableRowActions = /*@__PURE__*/ createReactComponent('modus-table-row-actions');
export const ModusTableRowActionsCell = /*@__PURE__*/ createReactComponent('modus-table-row-actions-cell');
export const ModusTableRowActionsMenu = /*@__PURE__*/ createReactComponent('modus-table-row-actions-menu');
export const ModusTableToolbar = /*@__PURE__*/ createReactComponent('modus-table-toolbar');
export const ModusTabs = /*@__PURE__*/ createReactComponent('modus-tabs');
export const ModusTextInput = /*@__PURE__*/ createReactComponent('modus-text-input');
export const ModusTextareaInput = /*@__PURE__*/ createReactComponent('modus-textarea-input');
export const ModusTimePicker = /*@__PURE__*/ createReactComponent('modus-time-picker');
export const ModusToast = /*@__PURE__*/ createReactComponent('modus-toast');
export const ModusToolbar = /*@__PURE__*/ createReactComponent('modus-toolbar');
export const ModusTooltip = /*@__PURE__*/ createReactComponent('modus-tooltip');
export const ModusTreeView = /*@__PURE__*/ createReactComponent('modus-tree-view');
export const ModusTreeViewItem = /*@__PURE__*/ createReactComponent('modus-tree-view-item');
export const ModusUtilityPanel = /*@__PURE__*/ createReactComponent('modus-utility-panel');
