// @trimble-oss/modus-web-components/dist/esm/modus-table.constants-c083d0c0.js
const SORTED_ASCENDING = 'Sorted Ascending';
const SORTED_DESCENDING = 'Sorted Descending';
const SORT_ASCENDING = 'Sort Ascending';
const SORT_DESCENDING = 'Sort Descending';
const HTML_ATTR_DATA_ACCESSOR_KEY = 'data-accessor-key';
const COLUMN_DEF_DATATYPE_KEY = 'dataType';
const COLUMN_DEF_SUB_ROWS_KEY = 'subRows';
const COLUMN_DEF_CELL_EDITOR_TYPE_KEY = 'cellEditorType';
const COLUMN_DEF_CELL_EDITOR_ARGS_KEY = 'cellEditorArgs';
const COLUMN_DEF_SHOWTOTAL = 'showTotal';
const COLUMN_DEF_ROW_SELECTION_ID = 'row-checkbox';
const COLUMN_DEF_ROW_SELECTION_CSS = 'row-checkbox';
const COLUMN_DEF_DATATYPE_TEXT = 'text';
const COLUMN_DEF_DATATYPE_INTEGER = 'integer';
const COLUMN_DEF_DATATYPE_LINK = 'link';
const COLUMN_DEF_DATATYPE_DATE = 'date';
const COLUMN_DEF_DATATYPE_BADGE = 'badge';
const CELL_EDIT_TYPE_TEXT = COLUMN_DEF_DATATYPE_TEXT;
const CELL_EDIT_TYPE_INT = COLUMN_DEF_DATATYPE_INTEGER;
const CELL_EDIT_TYPE_SELECT = 'select';
const CELL_EDIT_TYPE_AUTOCOMPLETE = 'autocomplete';
const CELL_EDIT_TYPE_DATE = COLUMN_DEF_DATATYPE_DATE;
const ALLOWED_CELL_EDIT_TYPES = [
    CELL_EDIT_TYPE_TEXT,
    CELL_EDIT_TYPE_INT,
    CELL_EDIT_TYPE_SELECT,
    CELL_EDIT_TYPE_AUTOCOMPLETE,
    CELL_EDIT_TYPE_DATE,
];
const PAGINATION_DEFAULT_SIZES = [10, 20, 30];
const PAGINATION_PAGEVIEW_TEXT = 'Page View';
const PAGINATION_SUMMARY_TEXT = 'Showing result';
// Keyboard key constants
const KEYBOARD_ENTER = 'enter';
const KEYBOARD_TAB = 'tab';
const KEYBOARD_LEFT = 'arrowleft';
const KEYBOARD_RIGHT = 'arrowright';
const KEYBOARD_UP = 'arrowup';
const KEYBOARD_DOWN = 'arrowdown';
const KEYBOARD_SPACE = ' ';
const KEYBOARD_ESCAPE = 'escape';
const EXPANDED_STATE_KEY = 'expanded';
const SORTING_STATE_KEY = 'sorting';
const ROW_SELECTION_STATE_KEY = 'rowSelection';
const COLUMN_SIZING_INFO_STATE_KEY = 'columnSizingInfo';
const PAGINATION_STATE_KEY = 'pagination';
const COLUMN_SIZING_STATE_KEY = 'columnSizing';
const COLUMN_VISIBILITY_STATE_KEY = 'columnVisibility';
const COLUMN_ORDER_STATE_KEY = 'columnOrder';

export { ALLOWED_CELL_EDIT_TYPES as A, COLUMN_DEF_DATATYPE_LINK as B, COLUMN_DEF_ROW_SELECTION_ID as C, COLUMN_DEF_CELL_EDITOR_TYPE_KEY as D, EXPANDED_STATE_KEY as E, CELL_EDIT_TYPE_INT as F, CELL_EDIT_TYPE_DATE as G, HTML_ATTR_DATA_ACCESSOR_KEY as H, CELL_EDIT_TYPE_AUTOCOMPLETE as I, CELL_EDIT_TYPE_SELECT as J, KEYBOARD_RIGHT as K, PAGINATION_SUMMARY_TEXT as P, ROW_SELECTION_STATE_KEY as R, SORTED_ASCENDING as S, PAGINATION_PAGEVIEW_TEXT as a, COLUMN_DEF_ROW_SELECTION_CSS as b, COLUMN_DEF_DATATYPE_KEY as c, COLUMN_DEF_DATATYPE_INTEGER as d, COLUMN_DEF_SHOWTOTAL as e, KEYBOARD_LEFT as f, KEYBOARD_ENTER as g, KEYBOARD_ESCAPE as h, KEYBOARD_TAB as i, COLUMN_DEF_SUB_ROWS_KEY as j, KEYBOARD_SPACE as k, SORTED_DESCENDING as l, SORT_ASCENDING as m, SORT_DESCENDING as n, PAGINATION_DEFAULT_SIZES as o, SORTING_STATE_KEY as p, PAGINATION_STATE_KEY as q, COLUMN_SIZING_STATE_KEY as r, COLUMN_VISIBILITY_STATE_KEY as s, COLUMN_ORDER_STATE_KEY as t, COLUMN_SIZING_INFO_STATE_KEY as u, KEYBOARD_UP as v, KEYBOARD_DOWN as w, COLUMN_DEF_CELL_EDITOR_ARGS_KEY as x, CELL_EDIT_TYPE_TEXT as y, COLUMN_DEF_DATATYPE_BADGE as z };
