// @trimble-oss/modus-web-components/dist/collection/components/modus-breadcrumb/modus-breadcrumb.js
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { h } from "@stencil/core";
export class ModusBreadcrumb {
    constructor() {
        this.ariaLabel = undefined;
        this.crumbs = [];
        this.underlineLinks = undefined;
    }
    render() {
        return (h("nav", { key: 'dd4973ca7376fe9f3a407c1b2b2a7df5aa4dee03', "aria-label": this.ariaLabel || undefined }, h("ol", { key: 'c4f4f45b97aaccef1d87ce26d9fa55621416d386' }, this.crumbs.map((crumb, index) => (h("li", { key: crumb.id }, index < this.crumbs.length - 1 ? (h("span", { class: `crumb ${this.underlineLinks ? 'underline' : ''}` }, h("a", { onClick: () => this.crumbClick.emit(crumb) }, crumb.display), h("span", { class: "divider" }, '>'))) : (h("span", { class: "last-crumb", "aria-current": "page" }, crumb.display))))))));
    }
    static get is() { return "modus-breadcrumb"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-breadcrumb.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-breadcrumb.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "string | null",
                    "resolved": "string",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The breadcrumb's aria-label."
                },
                "attribute": "aria-label",
                "reflect": false
            },
            "crumbs": {
                "type": "unknown",
                "mutable": false,
                "complexType": {
                    "original": "Crumb[]",
                    "resolved": "Crumb[]",
                    "references": {
                        "Crumb": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-breadcrumb/modus-breadcrumb.tsx",
                            "id": "src/components/modus-breadcrumb/modus-breadcrumb.tsx::Crumb"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "The breadcrumbs to render."
                },
                "defaultValue": "[]"
            },
            "underlineLinks": {
                "type": "boolean",
                "mutable": false,
                "complexType": {
                    "original": "boolean",
                    "resolved": "boolean",
                    "references": {}
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) A flag that controls the display of underline"
                },
                "attribute": "underline-links",
                "reflect": false
            }
        };
    }
    static get events() {
        return [{
                "method": "crumbClick",
                "name": "crumbClick",
                "bubbles": true,
                "cancelable": true,
                "composed": true,
                "docs": {
                    "tags": [],
                    "text": "(optional) An event that fires on breadcrumb click."
                },
                "complexType": {
                    "original": "Crumb",
                    "resolved": "Crumb",
                    "references": {
                        "Crumb": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-breadcrumb/modus-breadcrumb.tsx",
                            "id": "src/components/modus-breadcrumb/modus-breadcrumb.tsx::Crumb"
                        }
                    }
                }
            }];
    }
}
