// @trimble-oss/modus-web-components/dist/components/modus-table-columns-visibility2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import { i as KEYBOARD_TAB, g as KEYBOARD_ENTER, k as KEYBOARD_SPACE, B as KEYBOARD_DOWN, A as KEYBOARD_UP } from './modus-table.constants.js';
import { d as defineCustomElement$2 } from './modus-button2.js';
import { d as defineCustomElement$1 } from './modus-checkbox2.js';

const modusTableColumnsVisibilityCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.columns-visibility-menu{display:flex;flex-direction:column}.columns-visibility-menu .column-visibility-header{color:var(--modus-primary, #019aeb);font-size:0.875rem;font-weight:700;line-height:18px;margin:12px;margin-bottom:5px}.columns-visibility-menu .column-visibility-content{overflow:auto}.columns-visibility-menu .column-visibility-item{padding:5px 5px 5px 12px}.columns-visibility-menu .column-visibility-item:focus,.columns-visibility-menu .column-visibility-item:hover{background:var(--modus-group-item-hover-bg, #e0e1e9);outline:none}.columns-visibility-menu .column-visibility-item:last-child{margin-bottom:10px}.columns-visibility-menu .column-visibility-buttons-container{display:flex;justify-content:flex-end;margin-bottom:12px;margin-right:12px}modus-button{padding:12px 0 0 12px}";
const ModusTableColumnsVisibilityStyle0 = modusTableColumnsVisibilityCss;

const ModusTableColumnsVisibility = /*@__PURE__*/ proxyCustomElement(class ModusTableColumnsVisibility extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.refItemContent = [];
        this.getAllLeafColumns = undefined;
        this.columnsVisibility = undefined;
        this.showDropdown = undefined;
        this.menuIconContainerRef = undefined;
        this.toggleDropdown = undefined;
        this.maxHeight = undefined;
        this.columnsVisibilityState = new Map();
    }
    componentWillLoad() {
        // Initialize columns visibility state based on hiddenColumns
        this.initializeVisibilityState();
    }
    applyColumnsVisibility() {
        this.getAllLeafColumns().forEach((column) => {
            if (this.columnsVisibilityState.has(column.id)) {
                column.toggleVisibility(this.columnsVisibilityState.get(column.id));
            }
        });
        this.toggleDropdown(!this.showDropdown);
        this.menuIconContainerRef.focus();
    }
    closeDropdown() {
        this.columnsVisibilityState = new Map();
        this.toggleDropdown(!this.showDropdown);
        this.menuIconContainerRef.focus();
    }
    handleApplyKeyDown(event) {
        if (event.key.toLowerCase() === KEYBOARD_TAB && !event.shiftKey) {
            this.toggleDropdown(false);
        }
    }
    /**
     * Handling columns dropdown keyboard(arrowUp/arrowDown) navigation
     * @param event keyboard event
     * @param columnIndex column index
     */
    handleColumnItemKeyDown(event, columnIndex) {
        let currentRefItemIndex;
        const recursiveTillAnotherFocusItem = (refItemIndex, isIncrement) => {
            this.refItemContent[refItemIndex]
                ? this.refItemContent[refItemIndex].focus()
                : recursiveTillAnotherFocusItem(isIncrement ? refItemIndex + 1 : refItemIndex - 1, isIncrement);
        };
        const eventKey = event.key.toLowerCase();
        if (eventKey === KEYBOARD_ENTER || eventKey === KEYBOARD_SPACE) {
            this.toggleColumnVisibility(columnIndex);
        }
        else if (eventKey === KEYBOARD_DOWN) {
            currentRefItemIndex = columnIndex + 1 < this.refItemContent.length ? columnIndex + 1 : this.refItemContent.length - 1;
            recursiveTillAnotherFocusItem(currentRefItemIndex, true);
            event.preventDefault();
        }
        else if (eventKey === KEYBOARD_UP) {
            currentRefItemIndex = columnIndex - 1 >= 0 ? columnIndex - 1 : 0;
            recursiveTillAnotherFocusItem(currentRefItemIndex, false);
            event.preventDefault();
        }
    }
    /**
     * Adding column item refs(not disabled)
     */
    handleRefColumnItemContent(el, i, requiredColumn) {
        if (!requiredColumn) {
            this.refItemContent[i] = el;
        }
    }
    initializeVisibilityState() {
        // Set the visibility state based on hiddenColumns prop
        this.getAllLeafColumns().forEach((column) => {
            var _a, _b, _c, _d;
            const isRequired = (_b = (_a = this.columnsVisibility) === null || _a === void 0 ? void 0 : _a.requiredColumns) === null || _b === void 0 ? void 0 : _b.includes(column.id);
            const isHidden = (_d = (_c = this.columnsVisibility) === null || _c === void 0 ? void 0 : _c.hiddenColumns) === null || _d === void 0 ? void 0 : _d.includes(column.id);
            if (!isRequired && isHidden) {
                column.toggleVisibility(false);
            }
        });
    }
    toggleColumnVisibility(columnIndex) {
        const shadowRootChildren = this.refItemContent[columnIndex].children[0].shadowRoot
            .children;
        const modusCheckboxElement = shadowRootChildren[shadowRootChildren.length - 1];
        modusCheckboxElement.click();
    }
    /**
     * Returns true if the column is required column
     */
    checkIfRequiredColumn(columnId) {
        var _a, _b;
        return (_b = (_a = this.columnsVisibility) === null || _a === void 0 ? void 0 : _a.requiredColumns) === null || _b === void 0 ? void 0 : _b.includes(columnId);
    }
    renderColumnsChecklist() {
        const columnVisibilityItemControls = (columnId, columnIndex) => {
            return {
                key: columnId,
                tabIndex: this.checkIfRequiredColumn(columnId) ? -1 : 0,
                ref: (el) => {
                    this.handleRefColumnItemContent(el, columnIndex, this.checkIfRequiredColumn(columnId));
                },
                onClick: () => this.toggleColumnVisibility(columnIndex),
                onKeyDown: (event) => this.handleColumnItemKeyDown(event, columnIndex),
            };
        };
        return this.getAllLeafColumns().map((column, index) => {
            var _a, _b;
            return (h("div", Object.assign({}, columnVisibilityItemControls(column.id, index), { class: "column-visibility-item" }), h("modus-checkbox", { ariaLabel: column.columnDef.header, label: column.columnDef.header, checked: column.getIsVisible(), onCheckboxClick: (e) => this.columnsVisibilityState.set(column.id, e.detail), disabled: (_b = (_a = this.columnsVisibility) === null || _a === void 0 ? void 0 : _a.requiredColumns) === null || _b === void 0 ? void 0 : _b.includes(column.id), stopPropagation: true })));
        });
    }
    renderCancelApplyButtons() {
        return (h("div", { class: "column-visibility-buttons-container" }, h("modus-button", { size: "small", buttonStyle: "outline", onClick: () => this.closeDropdown() }, "Cancel"), h("modus-button", { size: "small", onClick: () => this.applyColumnsVisibility(), onKeyDown: (event) => this.handleApplyKeyDown(event) }, "Apply")));
    }
    render() {
        return (h("div", { key: '7357d1c0c3efb961c410ee2191f516673ff1fe31', class: "columns-visibility-menu", style: { maxHeight: this.maxHeight || '60dvh' } }, h("div", { key: '156fc3b9d58040a473f888d81f9e6a00021aac3b', class: "column-visibility-header" }, this.columnsVisibility.title || 'Columns'), h("div", { key: '588ed7b5939c4bced93b593032584c852208cd02', class: "column-visibility-content" }, this.renderColumnsChecklist()), this.renderCancelApplyButtons()));
    }
    static get style() { return ModusTableColumnsVisibilityStyle0; }
}, [1, "modus-table-columns-visibility", {
        "getAllLeafColumns": [16],
        "columnsVisibility": [16],
        "showDropdown": [4, "show-dropdown"],
        "menuIconContainerRef": [16],
        "toggleDropdown": [16],
        "maxHeight": [1, "max-height"],
        "columnsVisibilityState": [32]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-columns-visibility", "modus-button", "modus-checkbox"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-columns-visibility":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTableColumnsVisibility);
            }
            break;
        case "modus-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-checkbox":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusTableColumnsVisibility as M, defineCustomElement as d };
