// @trimble-oss/modus-web-components/dist/components/modus-table-row-actions-menu2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$2 } from './modus-list2.js';
import { d as defineCustomElement$1 } from './modus-list-item2.js';

const modusTableRowActionsMenuCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");modus-table-row-actions-menu .row-actions-menu{background-color:var(--modus-table-bg, #fff);box-shadow:0 0 2px rgba(37, 42, 46, 0.3);color:var(--modus-table-body-color, #252a2e);display:flex;position:fixed;top:-1px;z-index:9}modus-table-row-actions-menu modus-list-item{cursor:pointer}";
const ModusTableRowActionsMenuStyle0 = modusTableRowActionsMenuCss;

const ModusTableRowActionsMenu = /*@__PURE__*/ proxyCustomElement(class ModusTableRowActionsMenu extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        // a local state to keep track of the overflow icon click to prevent `handleClickOutside` from closing the menu
        // event.preventDefault would not solve the problem when multiple tables are present on the page
        this.isOverflowIconClicked = false;
        this.onOverflowRowActions = (e) => this.handleOverflowRowActions(e);
        this.onRowActionClick = (e) => this.handleRowActionButtonClick(e);
        this.onRowExpanded = () => (this.isMenuOpen = false);
        this.context = undefined;
        this.isMenuOpen = false;
        this.overFlowMenu = undefined;
        this.position = undefined;
        this.menuWidth = 0;
    }
    onMenuOpenChange(newValue) {
        if (!newValue) {
            this.overFlowMenu = null;
            this.position = null;
            this.tableRow = null;
        }
    }
    componentDidRender() {
        var _a;
        if (this.isMenuOpen) {
            const firstItem = (_a = Array.from(this.element.querySelectorAll('modus-list-item'))) === null || _a === void 0 ? void 0 : _a.find((el) => !el.disabled);
            firstItem === null || firstItem === void 0 ? void 0 : firstItem.focusItem();
            // Calculate and set the rowAction menu width
            const menuElement = this.element.querySelector('.row-actions-menu');
            if (menuElement) {
                this.menuWidth = menuElement.offsetWidth;
            }
        }
    }
    connectedCallback() {
        const { element } = this.context;
        element.addEventListener('overflowRowActions', this.onOverflowRowActions);
        element.addEventListener('rowActionClick', this.onRowActionClick);
        element.addEventListener('rowExpanded', this.onRowExpanded);
    }
    disconnectedCallback() {
        const { element } = this.context;
        element.removeEventListener('overflowRowActions', this.onOverflowRowActions);
        element.removeEventListener('rowActionClick', this.onRowActionClick);
        element.removeEventListener('rowExpanded', this.onRowExpanded);
    }
    handleOverflowRowActions(event) {
        var _a;
        const { componentId, actions, position, row, onClose } = event.detail;
        if (componentId !== this.context.componentId)
            return;
        this.isMenuOpen = ((_a = this.tableRow) === null || _a === void 0 ? void 0 : _a.id) === row.id ? false : true;
        if (this.isMenuOpen) {
            this.overFlowMenu = actions;
            this.position = position;
            this.tableRow = row;
            this.onCloseMenu = onClose;
        }
        this.isOverflowIconClicked = this.isMenuOpen;
        event.stopPropagation();
    }
    handleRowActionButtonClick({ detail: { actionId } }) {
        const rowActionButtonClicked = this.overFlowMenu && this.overFlowMenu.find((action) => action.id !== actionId);
        if (rowActionButtonClicked)
            this.isMenuOpen = false;
    }
    handleClickOutside(event) {
        if (!(this.element.contains(event.target) || this.isOverflowIconClicked)) {
            this.isMenuOpen = false;
        }
        this.isOverflowIconClicked = false;
    }
    handleListItemClick(id) {
        const { rowActionClick } = this.context;
        rowActionClick.emit({ actionId: id, row: this.tableRow.original });
    }
    handleListItemKeydown(e) {
        var _a;
        if (e.key.toLowerCase() === 'escape' || e.key.toLowerCase() === 'enter') {
            this.isMenuOpen = false;
            (_a = this.onCloseMenu) === null || _a === void 0 ? void 0 : _a.call(this);
            e.preventDefault();
        }
    }
    render() {
        var _a;
        if (!(((_a = this.overFlowMenu) === null || _a === void 0 ? void 0 : _a.length) && this.position))
            return null;
        const { x, y } = this.position;
        const style = {
            transform: `translate(calc(${x - this.menuWidth / 1.5}px - 8px), calc(${y}px))`,
        };
        return (h(Host, null, this.isMenuOpen && (h("div", { style: Object.assign({}, style), class: "row-actions-menu" }, h("modus-list", { class: "hydrated" }, this.overFlowMenu.map(({ label, id, isDisabled = () => false }) => {
            var _a;
            const disabled = isDisabled((_a = this.tableRow) === null || _a === void 0 ? void 0 : _a.original);
            return (h("modus-list-item", { style: { maxWidth: '155px' }, disabled: disabled, onItemClick: () => this.handleListItemClick(id), class: "hydrated row-actions-menu-item", onKeyDown: (e) => this.handleListItemKeydown(e), tabindex: disabled ? -1 : 0 }, label));
        }))))));
    }
    get element() { return this; }
    static get watchers() { return {
        "isMenuOpen": ["onMenuOpenChange"]
    }; }
    static get style() { return ModusTableRowActionsMenuStyle0; }
}, [0, "modus-table-row-actions-menu", {
        "context": [16],
        "isMenuOpen": [32],
        "overFlowMenu": [32],
        "position": [32],
        "menuWidth": [32]
    }, [[4, "click", "handleClickOutside"]], {
        "isMenuOpen": ["onMenuOpenChange"]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-row-actions-menu", "modus-list", "modus-list-item"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-row-actions-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTableRowActionsMenu);
            }
            break;
        case "modus-list":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-list-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusTableRowActionsMenu as M, defineCustomElement as d };
