// @trimble-oss/modus-web-components/dist/components/modus-table-filler-column2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const modusTableFillerColumnCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");modus-table-filler-column{z-index:0}modus-table-filler-column .d-none{display:none}modus-table-filler-column table thead tr,modus-table-filler-column table tfoot tr{height:inherit}modus-table-filler-column table:not(.cell-borderless) td:last-child,modus-table-filler-column table:not(.cell-borderless) th:last-child{border-right:none}modus-table-filler-column table:not(.cell-borderless) tbody td:first-child,modus-table-filler-column table:not(.cell-borderless) thead th:first-child{border-left:0.0625rem var(--modus-table-border-color, #b7b9c3) solid !important}";
const ModusTableFillerColumnStyle0 = modusTableFillerColumnCss;

const ModusTableFillerColumn = /*@__PURE__*/ proxyCustomElement(class ModusTableFillerColumn extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.observer = null;
        this.updateContainerLayout = () => {
            var _a, _b, _c, _d, _e, _f;
            const tableWidth = (_a = this.targetTable.getBoundingClientRect()) === null || _a === void 0 ? void 0 : _a.width;
            const parentWidth = (_c = (_b = this.targetTable.parentElement) === null || _b === void 0 ? void 0 : _b.getBoundingClientRect()) === null || _c === void 0 ? void 0 : _c.width;
            this.showFillerTable = tableWidth < parentWidth;
            if (!this.showFillerTable)
                return;
            if (this.fillerTableRef) {
                this.fillerTableRef.querySelector('thead').style.height = `${(_d = this.targetTable.querySelector('thead')) === null || _d === void 0 ? void 0 : _d.getBoundingClientRect().height}px`;
                if (this.summaryRow) {
                    this.fillerTableRef.querySelector('tfoot').style.height = `${(_e = this.targetTable.querySelector('tfoot')) === null || _e === void 0 ? void 0 : _e.getBoundingClientRect().height}px`;
                }
                this.fillerTableRef.querySelector('tbody').style.height = `${(_f = this.targetTable.querySelector('tbody')) === null || _f === void 0 ? void 0 : _f.getBoundingClientRect().height}px`;
            }
        };
        this.cellBorderless = undefined;
        this.summaryRow = undefined;
        this.container = undefined;
        this.showFillerTable = false;
    }
    componentDidLoad() {
        this.targetTable = this.container.shadowRoot.querySelector('table');
        if (this.targetTable) {
            this.updateContainerLayout();
            this.connectDOMObserver();
        }
    }
    disconnectedCallback() {
        this.disconnectDOMObserver();
    }
    connectDOMObserver() {
        this.observer = new ResizeObserver(this.updateContainerLayout);
        this.observer.observe(this.targetTable);
    }
    disconnectDOMObserver() {
        if (this.observer) {
            this.observer.disconnect();
        }
    }
    render() {
        return (h("table", { key: 'ee0a9f15c17f12f97c163b5a70d4cb0c629b7c76', id: "table-filler-column", class: { 'cell-borderless': this.cellBorderless, 'd-none': !this.showFillerTable }, ref: (el) => (this.fillerTableRef = el) }, h("thead", { key: 'bca2ba3f12619a0f7f3eaaba680dbebbc20a5309' }, h("tr", { key: '86e18debe16a7f8e08a59d742a6b2a0ba204e968' }, h("th", { key: '1c83623c16b1da4a35c4c6f0388052ef9f575caa' }))), h("tbody", { key: 'eb88f740a59f41e969238173557930a3e260be13' }, h("tr", { key: 'a84f32451ccc75f202abc17efbde0e782786f263' }, h("td", { key: 'e34fb7c702a0ba4b5c663a586c8096a96c85a7e8' }))), this.summaryRow && (h("tfoot", null, h("tr", { class: "summary-row" }, h("td", null))))));
    }
    static get style() { return ModusTableFillerColumnStyle0; }
}, [0, "modus-table-filler-column", {
        "cellBorderless": [4, "cell-borderless"],
        "summaryRow": [4, "summary-row"],
        "container": [16],
        "showFillerTable": [32]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-filler-column"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-filler-column":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTableFillerColumn);
            }
            break;
    } });
}

export { ModusTableFillerColumn as M, defineCustomElement as d };
