// @trimble-oss/modus-web-components/dist/components/modus-navbar-notifications-menu2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const modusNavbarNotificationsMenuCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.notifications-menu{background-color:var(--modus-body-bg, #fff);border:0.0625rem solid var(--modus-navbar-popup-border-color, #cbcdd6);border-radius:0.25rem;box-shadow:0 0 8px rgba(36, 35, 45, 0.3);color:var(--modus-body-color, #252a2e);display:flex;flex-direction:column;max-width:384px;padding:0;position:absolute;right:0;top:48px;z-index:100;background-color:var(--modus-navbar-popup-bg, #fff);cursor:default;width:max-content}.notifications-menu.reverse{left:0}";
const ModusNavbarNotificationsMenuStyle0 = modusNavbarNotificationsMenuCss;

const ModusNavbarNotificationsMenu = /*@__PURE__*/ proxyCustomElement(class ModusNavbarNotificationsMenu extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.reverse = undefined;
    }
    render() {
        const direction = this.reverse ? 'reverse' : '';
        return (h("div", { key: '531acef768a44cdbe68018635b96512adb05becf', class: `notifications-menu ${direction}`, onClick: (event) => event.preventDefault() }, h("slot", { key: 'c8b94931af47161572850fff2d104b0a70307d57' })));
    }
    static get style() { return ModusNavbarNotificationsMenuStyle0; }
}, [1, "modus-navbar-notifications-menu", {
        "reverse": [4]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-navbar-notifications-menu"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-navbar-notifications-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusNavbarNotificationsMenu);
            }
            break;
    } });
}

export { ModusNavbarNotificationsMenu as M, defineCustomElement as d };
