// @trimble-oss/modus-web-components/dist/components/modus-table-toolbar2.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';
import { d as defineCustomElement$4 } from './modus-button2.js';
import { d as defineCustomElement$3 } from './modus-checkbox2.js';
import { d as defineCustomElement$2 } from './modus-table-columns-visibility2.js';
import { d as defineCustomElement$1 } from './modus-table-dropdown-menu2.js';

const modusTableToolbarCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.table-toolbar{align-items:center;display:flex;justify-content:space-between;padding-bottom:3px}.table-toolbar .section{align-items:center;display:flex}";
const ModusTableToolbarStyle0 = modusTableToolbarCss;

const ModusTablePanel = /*@__PURE__*/ proxyCustomElement(class ModusTablePanel extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.context = undefined;
    }
    render() {
        // const { tableInstance: table, toolbarOptions: options } = this.context;
        return (h(Host, { key: 'fcbd8a1bec08ca5f4a11de68eac6b2a708127ee1' }, h("div", { key: 'd218fafadd7beb53e8c43ab0e45bf566a5996a0b', class: "table-toolbar" }, h("div", { key: 'e1d07736405cbb3828070c2f9b93a87fb9848937', class: "section" }, h("slot", { key: '89737c466d0a8e0baea9a4e9f89bf64893bc110c', name: "group-left" })), h("div", { key: '86342197962d9c921c8b8421539cee22a36226f9', class: "section" }, h("slot", { key: 'eb7745a45dc1c629d174414004efb47803971b49', name: "group-right" }), h("modus-table-dropdown-menu", { context: this.context })))));
    }
    static get style() { return ModusTableToolbarStyle0; }
}, [1, "modus-table-toolbar", {
        "context": [16]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-table-toolbar", "modus-button", "modus-checkbox", "modus-table-columns-visibility", "modus-table-dropdown-menu"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-table-toolbar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusTablePanel);
            }
            break;
        case "modus-button":
            if (!customElements.get(tagName)) {
                defineCustomElement$4();
            }
            break;
        case "modus-checkbox":
            if (!customElements.get(tagName)) {
                defineCustomElement$3();
            }
            break;
        case "modus-table-columns-visibility":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-table-dropdown-menu":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusTablePanel as M, defineCustomElement as d };
