// @trimble-oss/modus-web-components/dist/components/modus-divider2.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';

const modusDividerCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.modus-divider{border-left:1px solid var(--modus-toolbar-divider-bg, #252a2e);height:32px;margin-left:3px;margin-right:3px;opacity:0.25}";
const ModusDividerStyle0 = modusDividerCss;

const ModusDivider = /*@__PURE__*/ proxyCustomElement(class ModusDivider extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
    }
    render() {
        return h("div", { key: 'a8e995a07024a15445fa467aad5a0f48f50fefb8', class: "modus-divider" });
    }
    static get style() { return ModusDividerStyle0; }
}, [0, "modus-divider"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-divider"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-divider":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusDivider);
            }
            break;
    } });
}

export { ModusDivider as M, defineCustomElement as d };
