// @trimble-oss/modus-web-components/dist/components/modus-navbar-profile-menu2.js
import { proxyCustomElement, HTMLElement, createEvent, h } from '@stencil/core/internal/client';
import { M as ModusIconMap } from './ModusIconMap.js';
import { d as defineCustomElement$2 } from './modus-list2.js';
import { d as defineCustomElement$1 } from './modus-list-item2.js';

const modusNavbarProfileMenuCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.profile-menu{background-color:var(--modus-body-bg, #fff);border:0.0625rem solid var(--modus-navbar-popup-border-color, #cbcdd6);border-radius:0.25rem;box-shadow:0 0 8px rgba(36, 35, 45, 0.3);color:var(--modus-body-color, #252a2e);display:flex;flex-direction:column;max-width:384px;padding:0;position:absolute;right:0;top:48px;z-index:100;align-items:center;background-color:var(--modus-navbar-popup-bg, #fff);cursor:default;padding:0;width:19rem}.profile-menu .user{align-items:center;display:flex;flex-direction:column;justify-content:center;padding:1rem 1.5rem 0 1.5rem}.profile-menu .user .avatar,.profile-menu .user .initials{background-color:var(--modus-navbar-icon-color, #252a2e);border-radius:50%;height:6rem;margin-bottom:1rem;user-select:none;width:6rem}.profile-menu .user .initials{align-items:center;color:var(--modus-navbar-profile-icon-initials-color, #fff);display:flex;font-size:1.5rem;font-weight:600;justify-content:center}.profile-menu .user .user-details{align-items:center;display:flex;flex-direction:column;justify-content:center;margin-bottom:1rem;max-width:16rem}.profile-menu .user .user-details .username{font-size:1rem;font-weight:600;overflow-wrap:break-word;text-align:center;width:fit-content}.profile-menu .user .user-details .email{font-size:0.75rem;overflow-wrap:anywhere;text-align:center}.profile-menu.profile-menu-blue .user .avatar,.profile-menu.profile-menu-blue .user .initials{background-color:#b7b9c3;color:var(--modus-navbar-blue-profile-icon-initials-color, #0e416c)}.profile-menu .links{cursor:pointer;margin-bottom:0.75rem;max-width:19rem;user-select:none;width:100%}.profile-menu .links .link-item{align-items:center;display:flex;flex-direction:row;white-space:normal;word-wrap:break-word}.profile-menu .links .icon{align-items:center;display:flex;justify-content:center;margin-right:1rem}.profile-menu .sign-out{background-color:var(--modus-navbar-profile-menu-sign-out-background-color, #f1f1f6);user-select:none;width:100%}.profile-menu .sign-out div{color:var(--modus-blue, #0063a3);cursor:pointer;font-size:0.875rem;padding:0.625rem 1rem}@media (hover: hover){.profile-menu .sign-out div:hover{text-decoration:underline}}.profile-menu.reverse{left:0}";
const ModusNavbarProfileMenuStyle0 = modusNavbarProfileMenuCss;

const ModusNavbarProfileMenu = /*@__PURE__*/ proxyCustomElement(class ModusNavbarProfileMenu extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.linkClick = createEvent(this, "linkClick", 7);
        this.signOutClick = createEvent(this, "signOutClick", 7);
        this.avatarUrl = undefined;
        this.email = undefined;
        this.initials = undefined;
        this.links = undefined;
        this.signOutText = 'Sign out';
        this.reverse = undefined;
        this.username = undefined;
        this.variant = 'default';
    }
    signOutKeydownHandler(event) {
        if (event.code !== 'Enter') {
            return;
        }
        this.signOutClick.emit();
    }
    render() {
        var _a;
        const direction = this.reverse ? 'reverse' : '';
        const variant = this.variant === 'default' ? '' : `profile-menu-${this.variant}`;
        return (h("div", { key: '950336bb8022523f5d9d82b680b1f93627394c66', class: `profile-menu ${direction} ${variant}`, onClick: (event) => event.preventDefault() }, h("div", { key: '3ade3406c483372cdd6ff51e8ec076c3de2a0abe', class: "user" }, this.avatarUrl ? (h("img", { class: "avatar", src: this.avatarUrl, alt: "Avatar" })) : (h("span", { class: "initials" }, this.initials)), h("div", { key: '3fab86cf79860d4d8a450d67657747a2ff980469', class: "user-details" }, h("div", { key: 'c6bbb2a6a57ec589696042009fc681907852e693', class: "username" }, this.username), h("div", { key: 'e7badf44abd0124cd3e3ff7632e878212a9bdb97', class: "email" }, this.email))), ((_a = this.links) === null || _a === void 0 ? void 0 : _a.length) ? (h("modus-list", { class: "links" }, this.links.map((link) => {
            return (h("modus-list-item", { borderless: true, onItemClick: () => this.linkClick.emit(link.id) }, h("div", { class: "link-item" }, link.icon && (h("div", { class: "icon" }, h(ModusIconMap, { icon: link.icon, size: "24" }))), link.display)));
        }))) : null, h("slot", { key: '4056c5fcd4b8180ce1db6424b4b6a28657ba3497' }), h("div", { key: '85a779b160a7cba9e63053d122a363e70e19f721', class: "sign-out", onClick: () => this.signOutClick.emit(), onKeyDown: (event) => this.signOutKeydownHandler(event), tabIndex: 0 }, h("div", { key: '1d7400ca1ae72f7b056f93ac95daf283461badd5' }, this.signOutText))));
    }
    static get style() { return ModusNavbarProfileMenuStyle0; }
}, [1, "modus-navbar-profile-menu", {
        "avatarUrl": [1, "avatar-url"],
        "email": [1],
        "initials": [1],
        "links": [16],
        "signOutText": [1, "sign-out-text"],
        "reverse": [4],
        "username": [1],
        "variant": [1]
    }]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-navbar-profile-menu", "modus-list", "modus-list-item"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-navbar-profile-menu":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusNavbarProfileMenu);
            }
            break;
        case "modus-list":
            if (!customElements.get(tagName)) {
                defineCustomElement$2();
            }
            break;
        case "modus-list-item":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusNavbarProfileMenu as M, defineCustomElement as d };
