// @trimble-oss/modus-web-components/dist/components/modus-progress-bar.js
import { proxyCustomElement, HTMLElement, h, Host } from '@stencil/core/internal/client';

const modusProgressBarCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");.modus-progress-bar{display:flex;font-family:\"Open Sans\", -apple-system, BlinkMacSystemFont, \"Segoe UI\", Roboto, \"Helvetica Neue\", Arial, \"Noto Sans\", sans-serif, \"Apple Color Emoji\", \"Segoe UI Emoji\", \"Segoe UI Symbol\", \"Noto Color Emoji\";height:1rem;outline:0.0625rem solid var(--modus-progress-border-color, #6a6e79);width:100%}.modus-progress-bar.default-background-color{background-color:var(--modus-progress-bg, #fff)}.modus-progress-bar .progress{display:flex;font-size:0.75rem;forced-color-adjust:none;justify-content:center;overflow:hidden;transition:width 0.6s ease}.modus-progress-bar .progress.default-color{background-color:var(--modus-progress-bar-bg, #0063a3)}.modus-progress-bar .progress.default-text-color{color:#fff}.modus-progress-bar.small{height:0.5rem}.modus-progress-bar.compact{height:0.25rem}";
const ModusProgressBarStyle0 = modusProgressBarCss;

const ModusProgressBar$1 = /*@__PURE__*/ proxyCustomElement(class ModusProgressBar extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.__attachShadow();
        this.classBySize = new Map([
            ['default', 'default'],
            ['small', 'small'],
            ['compact', 'compact'],
        ]);
        this.ariaLabel = undefined;
        this.backgroundColor = undefined;
        this.color = undefined;
        this.maxValue = 100;
        this.minValue = 0;
        this.size = 'default';
        this.text = undefined;
        this.textColor = undefined;
        this.value = 0;
    }
    getProgressStyle(percentage) {
        const progressStyle = {
            backgroundColor: this.color,
            color: this.textColor,
            width: `${percentage}%`,
        };
        return progressStyle;
    }
    getProgressBarStyle() {
        const progressBarStyle = {
            backgroundColor: this.backgroundColor,
        };
        return progressBarStyle;
    }
    render() {
        const percentage = ((this.value - this.minValue) / (this.maxValue - this.minValue)) * 100;
        const progressBarBackgroundColorClass = this.backgroundColor ? '' : 'default-background-color';
        const progressColorClass = this.color ? '' : 'default-color';
        const progressTextColor = this.textColor ? '' : 'default-text-color';
        const progressBarClass = `
      modus-progress-bar
      ${progressBarBackgroundColorClass}
      ${this.classBySize.get(this.size)}
     `;
        const progressClass = `progress ${progressColorClass} ${progressTextColor}`;
        return (h(Host, { key: '15aac189d0c000836c9283530a63590d1348b0e0', "aria-label": this.ariaLabel, "aria-valuemax": this.maxValue, "aria-valuemin": this.minValue, "aria-valuenow": this.value, role: "progressbar" }, h("div", { key: 'f1ad64202db18e70ca42b828d9ac8d6e80852426', class: progressBarClass, style: this.getProgressBarStyle() }, h("div", { key: '65497e159fce3bfe93481f0ab069c8a5b34bf57f', class: progressClass, style: this.getProgressStyle(percentage) }, this.size === 'default' && this.text))));
    }
    static get style() { return ModusProgressBarStyle0; }
}, [1, "modus-progress-bar", {
        "ariaLabel": [1, "aria-label"],
        "backgroundColor": [1, "background-color"],
        "color": [1],
        "maxValue": [2, "max-value"],
        "minValue": [2, "min-value"],
        "size": [1],
        "text": [1],
        "textColor": [1, "text-color"],
        "value": [2]
    }]);
function defineCustomElement$1() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-progress-bar"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-progress-bar":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusProgressBar$1);
            }
            break;
    } });
}

const ModusProgressBar = ModusProgressBar$1;
const defineCustomElement = defineCustomElement$1;

export { ModusProgressBar, defineCustomElement };
