// @trimble-oss/modus-web-components/dist/components/modus-navbar-search-overlay2.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { I as IconClose } from './icon-close.js';
import { d as defineCustomElement$1 } from './modus-text-input2.js';

const modusNavbarSearchOverlayCss = "@import url(\"https://fonts.googleapis.com/css?family=Open+Sans:300,400,600,700\");:host{display:block}.overlay-wrapper{display:flex;flex-shrink:0;gap:45px}.overlay-wrapper .search-box{display:flex;flex-grow:2;justify-content:center}.overlay-wrapper .search-box modus-text-input{display:block;margin-left:1rem;max-width:764px;min-width:185px;width:100%}.overlay-wrapper .navbar-button{justify-content:end;padding-right:1rem}@media screen and (width <= 576px){.overlay-wrapper .navbar-button{padding-right:0.5rem}}";
const ModusNavbarSearchOverlayStyle0 = modusNavbarSearchOverlayCss;

const ModusNavbarSearchOverlay = /*@__PURE__*/ proxyCustomElement(class ModusNavbarSearchOverlay extends HTMLElement {
    constructor() {
        super();
        this.__registerHost();
        this.close = createEvent(this, "close", 7);
        this.search = createEvent(this, "search", 7);
    }
    componentDidRender() {
        var _a;
        (_a = this.modusTextInput) === null || _a === void 0 ? void 0 : _a.focusInput();
    }
    render() {
        return (h(Host, { key: 'e7e2f6c0a27303b1e13ee3ba5636408e074a6686' }, h("div", { key: 'dc742d0a964a7e1ca15dafb931b42fd322abc915', class: "overlay-wrapper" }, h("div", { key: 'c2038d34335710412d41aa8e84955ad9b2cec947', class: "search-box" }, h("modus-text-input", { key: '6ff64e9b3537ca1fdca5ae8807cc5ee8b0237558', placeholder: "Search", size: "large", type: "search", clearable: true, onValueChange: (event) => this.search.emit(event.detail), "include-search-icon": true, ref: (el) => (this.modusTextInput = el) })), h("div", { key: 'aeab44ccf9ac551b03c8487dcd0b718a673bdf65', class: "navbar-button", "data-test-id": "close-button" }, h("span", { key: 'c841a0b80c58edbb2dc2237a730df3a3859ffd8f', class: "navbar-button-icon", role: "button", tabIndex: 0, onKeyDown: (event) => event.code === 'Enter' && this.close.emit(), onClick: () => this.close.emit() }, h(IconClose, { key: '34697a8894bd5ac23eed9ce45147cd2825635bbc', size: "24" }))))));
    }
    get element() { return this; }
    static get style() { return ModusNavbarSearchOverlayStyle0; }
}, [0, "modus-navbar-search-overlay"]);
function defineCustomElement() {
    if (typeof customElements === "undefined") {
        return;
    }
    const components = ["modus-navbar-search-overlay", "modus-text-input"];
    components.forEach(tagName => { switch (tagName) {
        case "modus-navbar-search-overlay":
            if (!customElements.get(tagName)) {
                customElements.define(tagName, ModusNavbarSearchOverlay);
            }
            break;
        case "modus-text-input":
            if (!customElements.get(tagName)) {
                defineCustomElement$1();
            }
            break;
    } });
}

export { ModusNavbarSearchOverlay as M, defineCustomElement as d };
