// @trimble-oss/modus-web-components/dist/collection/components/modus-badge/modus-badge.js
// eslint-disable-next-line
import { h } from "@stencil/core";
export class ModusBadge {
    constructor() {
        this.classByColor = new Map([
            ['danger', 'color-danger'],
            ['dark', 'color-dark'],
            ['primary', 'color-primary'],
            ['secondary', 'color-secondary'],
            ['success', 'color-success'],
            ['tertiary', 'color-tertiary'],
            ['warning', 'color-warning'],
        ]);
        this.classBySize = new Map([
            ['small', 'size-small'],
            ['medium', 'size-medium'],
            ['large', 'size-large'],
        ]);
        this.classByType = new Map([
            ['counter', 'type-counter'],
            ['default', 'type-default'],
            ['text', 'type-text'],
        ]);
        this.ariaLabel = undefined;
        this.color = 'primary';
        this.size = 'medium';
        this.type = 'default';
    }
    render() {
        const className = `badge ${this.classByColor.get(this.color)} ${this.classBySize.get(this.size)} ${this.classByType.get(this.type)}`;
        return (h("div", { key: '3d67a735e88171cdb16e4d2b3254fed8532fa039', "aria-label": this.ariaLabel || undefined, class: className, role: "status" }, h("slot", { key: 'e60c8c096a0f4e36e60b3dd5f5d04ce495d2fe65' })));
    }
    static get is() { return "modus-badge"; }
    static get encapsulation() { return "shadow"; }
    static get originalStyleUrls() {
        return {
            "$": ["modus-badge.scss"]
        };
    }
    static get styleUrls() {
        return {
            "$": ["modus-badge.css"]
        };
    }
    static get properties() {
        return {
            "ariaLabel": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeProperties['ariaLabel']",
                    "resolved": "string",
                    "references": {
                        "BadgeProperties": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-badge/modus-badge.tsx",
                            "id": "src/components/modus-badge/modus-badge.tsx::BadgeProperties"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The badge's aria-label"
                },
                "attribute": "aria-label",
                "reflect": false
            },
            "color": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeProperties['color']",
                    "resolved": "\"danger\" | \"dark\" | \"primary\" | \"secondary\" | \"success\" | \"tertiary\" | \"warning\"",
                    "references": {
                        "BadgeProperties": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-badge/modus-badge.tsx",
                            "id": "src/components/modus-badge/modus-badge.tsx::BadgeProperties"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The color of the badge"
                },
                "attribute": "color",
                "reflect": false,
                "defaultValue": "'primary'"
            },
            "size": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeProperties['size']",
                    "resolved": "\"large\" | \"medium\" | \"small\"",
                    "references": {
                        "BadgeProperties": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-badge/modus-badge.tsx",
                            "id": "src/components/modus-badge/modus-badge.tsx::BadgeProperties"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The size of the badge"
                },
                "attribute": "size",
                "reflect": false,
                "defaultValue": "'medium'"
            },
            "type": {
                "type": "string",
                "mutable": false,
                "complexType": {
                    "original": "BadgeProperties['type']",
                    "resolved": "\"counter\" | \"default\" | \"text\"",
                    "references": {
                        "BadgeProperties": {
                            "location": "local",
                            "path": "/home/runner/work/modus-web-components/modus-web-components/stencil-workspace/src/components/modus-badge/modus-badge.tsx",
                            "id": "src/components/modus-badge/modus-badge.tsx::BadgeProperties"
                        }
                    }
                },
                "required": false,
                "optional": false,
                "docs": {
                    "tags": [],
                    "text": "(optional) The type of the badge"
                },
                "attribute": "type",
                "reflect": false,
                "defaultValue": "'default'"
            }
        };
    }
}
